### PR TITLE
KAFKA-10326 Both serializer and deserializer should be able to see th…

### PR DESCRIPTION
**Pull request by chia7712**
https://issues.apache.org/jira/browse/KAFKA-10326

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by chia7712** [comment]
```EosBetaUpgradeIntegrationTest``` is flaky... retest this please

**Review by abbccdda — COMMENTED** [review]
Are there any serde use cases that require client id?

**Comment by chia7712** [comment]
> Are there any serde use cases that require client id?

I observed this issue when we are removing custom client id config. It causes error as the metrics of our serde is encoded with client id.

File: clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java [hunk 1]
```diff
@@ -228,6 +228,13 @@ public Password getPassword(String key) {
         return copy;
     }
 
+    public Map<String, Object> originals(Map<String, Object> configOverrides) {
```
Comment (abbccdda) [inline clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java:231]: s/originals/copyWithOverride?

**Review by abbccdda — COMMENTED** [review]
The change makes sense to me. Will also let @guozhangwang take a look. Btw, I don't think this is an improvement rather than a bug, as we don't have any guarantee to see client id in serdes before.

**Comment by chia7712** [comment]
> Btw, I don't think this is an improvement rather than a bug, as we don't have any guarantee to see client id in serdes before.

There is another reason. We do pass generated client id to metric reporter.  It seems to me all plugins should see consistent props.

File: clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java (hunk 1 above)
Comment (chia7712) [reply clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java:231]: Will copy that

**Comment by guozhangwang** [comment]
This change makes sense to me, thanks @chia7712 

One (very) minor thing is that, `enableAutoCommit` may also be overridden when passing into the coordinator object, i.e. the value read from ConsumerConfig may be different from what's actually used. I'm not too worry that this would be leveraged by serde but just want to bring this up.

**Comment by chia7712** [comment]
@guozhangwang thanks for reviews

> One (very) minor thing is that,  may also be overridden when passing into the coordinator object, i.e. the value read from ConsumerConfig may be different from what's actually used. I'm not too worry that this would be leveraged by serde but just want to bring this up.

Let me address it to make consistent configs for all plugins.

File: clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java (hunk 1 above)
Comment (chia7712) [reply clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java:231]: I have updated the method name.

**Comment by chia7712** [comment]
```org.apache.kafka.clients.admin.KafkaAdminClientTest.testMetadataRetries``` is traced by #9091 and ```EosBetaUpgradeIntegrationTest.shouldUpgradeFromEosAlphaToEosBeta[true]``` is flaky. retest this please

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 2]
```diff
@@ -2511,4 +2511,39 @@ public void testEnforceRebalanceTriggersRebalanceOnNextPoll() {
...
+        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, DeserializerForClientId.class.getName());
+
+        KafkaConsumer<byte[], byte[]> consumer = new KafkaConsumer<>(props);
+
+        assertEquals(2, DeserializerForClientId.CLIENT_IDS.size());
+        assertEquals(DeserializerForClientId.CLIENT_IDS.get(0), consumer.getClientId());
+        assertEquals(DeserializerForClientId.CLIENT_IDS.get(1), consumer.getClientId());
+
+        assertEquals(2, DeserializerForClientId.AUTO_COMMIT_FLAGS.size());
+        assertEquals(DeserializerForClientId.AUTO_COMMIT_FLAGS.get(0), consumer.isAutoCommitEnabled());
```
Comment (abbccdda) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: This seems beyond the scope of this PR. Why do we also need to let deserializer see whether given consumer instance is auto commit enabled?

Comment (chia7712) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: This flag is included by reason of this comment https://github.com/apache/kafka/pull/9102#issuecomment-667734629

I’m ok to revert it if it is overkill

**Comment by chia7712** [comment]
@guozhangwang @abbccdda More comments?

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java (hunk 2 above)
Comment (abbccdda) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: Sorry for the delay, I suggest we revert this part to properly scope this PR.

Comment (chia7712) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: @abbccdda thanks for suggestion. I have reverted it.

**Comment by chia7712** [comment]
@guozhangwang @abbccdda ping ~

**Comment by guozhangwang** [comment]
I took a quick look and it lgtm. @abbccdda If your pass is good please feel free to merge.

**Comment by chia7712** [comment]
@abbccdda ping~

**Comment by chia7712** [comment]
rebase to trigger new QA

**Comment by guozhangwang** [comment]
I've merged it to trunk and marked fix version as 2.7, thanks a lot for your contribution @chia7712 !

**Pull request merged**